haze/src/androidMain/kotlin/dev/chrisbanes/haze/RenderScriptBlurEffect.kt

```diff
@@ -0,0 +1,183 @@
+// Copyright 2025, Christopher Banes and the Haze project contributors
+// SPDX-License-Identifier: Apache-2.0
+
+package dev.chrisbanes.haze
+
+import android.content.Context
+import android.graphics.Color
+import android.graphics.PorterDuff
+import android.os.Build
+import android.view.Surface
+import androidx.compose.ui.geometry.Size
+import androidx.compose.ui.geometry.toRect
+import androidx.compose.ui.graphics.Canvas
+import androidx.compose.ui.graphics.GraphicsContext
+import androidx.compose.ui.graphics.asImageBitmap
+import androidx.compose.ui.graphics.drawscope.CanvasDrawScope
+import androidx.compose.ui.graphics.drawscope.DrawScope
+import androidx.compose.ui.graphics.layer.GraphicsLayer
+import androidx.compose.ui.graphics.layer.drawLayer
+import androidx.compose.ui.graphics.withSaveLayer
+import androidx.compose.ui.node.currentValueOf
+import androidx.compose.ui.node.requireDensity
+import androidx.compose.ui.platform.LocalContext
+import androidx.compose.ui.platform.LocalGraphicsContext
+import androidx.compose.ui.platform.LocalLayoutDirection
+import androidx.compose.ui.unit.Density
+import androidx.compose.ui.unit.IntSize
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+
+internal class RenderScriptBlurEffect(
+  private val node: HazeEffectNode,
+) : BlurEffect {
+  private var renderScriptContext: RenderScriptContext? = null
+  private val drawScope = CanvasDrawScope()
+
+  private var drawGraphicsLayer: GraphicsLayer? = null
+  private val graphicsContext: GraphicsContext
+    get() = node.currentValueOf(LocalGraphicsContext)
+
+  private val density: Density
+    get() = node.requireDensity()
+
+  override fun DrawScope.drawEffect() {
+    val offset = node.layerOffset
+    var scaleFactor = node.calculateInputScaleFactor()
+    var blurRadiusPx = scaleFactor * with(density) { node.resolveBlurRadius().toPx() }
+
+    if (blurRadiusPx > MAX_BLUR_RADIUS) {
+      // RenderScript has a max blur radius (25px), so to create an equivalent visual effect
+      // we need to increase the scale factor
+      scaleFactor *= (MAX_BLUR_RADIUS / blurRadiusPx)
+      blurRadiusPx = MAX_BLUR_RADIUS
+    }
+
+    HazeLogger.d(TAG) { "drawEffect. blurRadius=${blurRadiusPx}px. scaleFactor=$scaleFactor" }
+
+    val scaledLayer = createScaledContentLayer(
+      node = node,
+      scaleFactor = scaleFactor,
+      layerSize = node.layerSize,
+      layerOffset = offset,
+    )
+    if (scaledLayer != null) {
+      node.coroutineScope.launch(Dispatchers.Main.immediate) {
+        updateSurface(content = scaledLayer, blurRadius = blurRadiusPx)
+      }
+    }
+
+    drawGraphicsLayer?.let { layer ->
+      drawScaledContentLayer(offset = -offset, scaleFactor = scaleFactor) {
+        // TODO: apply mask or progressive to layer?
+        // TODO: draw noise?
```

Comment:
Did you plan to draw the noise and progressive mask using the GraphicsLayer API, then blend it with the blurred bitmap, or were you considering doing this directly via RenderScript to produce the final bitmap? If you haven’t started on the latter approach yet, I might give it a try 🧐

Reply:
I was planning on just using the Canvas APIs as a first pass: `Paint` + mask shader  + `saveLayer`